packages/libs/deer/src/lib.rs

```diff
@@ -1,9 +1,451 @@
-//! Intentionally left blank for now!
-
 #![cfg_attr(not(feature = "std"), no_std)]
-#![warn(missing_docs, unreachable_pub, clippy::pedantic, clippy::nursery)]
+#![warn(unreachable_pub, clippy::pedantic, clippy::nursery)]
+// TODO: once more stable introduce: warning missing_docs
 #![allow(clippy::redundant_pub_crate)]
 #![allow(clippy::module_name_repetitions)]
 #![forbid(unsafe_code)]
 
+use alloc::{
+    collections::BTreeMap,
+    string::{String, ToString},
+    vec::Vec,
+};
+use core::marker::PhantomData;
+
+use error_stack::{IntoReport, Report, Result, ResultExt};
+use num_traits::ToPrimitive;
+
+pub use crate::{error::Error, number::Number};
+
+mod error;
+mod number;
+
 extern crate alloc;
+
+pub trait ObjectAccess<'de> {
+    type Error: Error;
+
+    fn value<T>(&mut self, key: &str) -> Result<T, Self::Error>
+    where
+        T: Deserialize<'de>;
+
+    fn next<T>(&mut self) -> Result<Option<(String, T)>, Self::Error>
+    where
+        T: Deserialize<'de>;
+
+    fn finish(self) -> Result<(), Self::Error>;
+}
+
+pub trait ArrayAccess<'de> {
+    type Error: Error;
+
+    fn next<T>(&mut self) -> Result<Option<T>, Self::Error>
+    where
+        T: Deserialize<'de>;
+
+    fn finish(self) -> Result<(), Self::Error>;
+}
+
+// TODO: Error PR: attach the expected and received type
+pub trait Visitor<'de>: Sized {
+    type Error: Error;
+    type Value;
+
+    fn expecting_display(&self) -> String;
+    // TODO: this is currently completely untyped, we might want to adhere to a standard, like
+    //  JSON-Schema or OpenAPI
+    //  The problem here mainly is: which crate to use, one can use utoipa (but that has significant
+    //  overhead)  there's no real library out there that properly just provides they types
```

Comment:
```suggestion
    //  overhead)  there's no real library out there that properly just provides the types
```

Unsure if this is what you meant